modyn/tests/supervisor/internal/pipeline_executor/test_pipeline_executor.py

```diff
@@ -356,30 +356,96 @@ def test__handle_triggers_within_batch_empty_triggers(
     trigger_ids = [(0, {}), (1, {}), (2, {})]
     test_inform_selector_and_trigger.side_effect = trigger_ids
     test_inform_selector.return_value = {}
-    test_get_number_of_samples.return_value = len(batch)
-
+    test_get_number_of_samples.side_effect = [0, 0, 4]
     pe._handle_triggers_within_batch(batch, triggering_indices)
-
     inform_selector_and_trigger_expected_args = [
         call(42, []),
         call(42, []),
-        call(42, [(10, 1), (11, 2), (12, 3), (13, 4)]),
+        call(42, [(10, 1, 5), (11, 2, 5), (12, 3, 5), (13, 4, 5)]),
     ]
     assert test_inform_selector_and_trigger.call_count == 3
     assert test_inform_selector_and_trigger.call_args_list == inform_selector_and_trigger_expected_args
 
-    run_training_expected_args = [call(0), call(1), call(2)]
-    assert test__run_training.call_count == 3
+    run_training_expected_args = [call(2)]
+    assert test__run_training.call_count == 1
     assert test__run_training.call_args_list == run_training_expected_args
 
     assert test_inform_selector.call_count == 1
-    test_inform_selector.assert_called_once_with(42, [(14, 5), (15, 6), (16, 7)])
+    test_inform_selector.assert_called_once_with(42, [(14, 5, 5), (15, 6, 5), (16, 7, 5)])
+
+
+@patch.object(PipelineExecutor, "_run_training")
+@patch.object(GRPCHandler, "inform_selector_and_trigger")
+@patch.object(GRPCHandler, "inform_selector")
+@patch.object(GRPCHandler, "get_number_of_samples")
+def test__handle_triggers_within_batch_trigger_timespan(
+    test_get_number_of_samples: MagicMock,
+    test_inform_selector: MagicMock,
+    test_inform_selector_and_trigger: MagicMock,
+    test__run_training: MagicMock,
+):
+    pe = get_non_connecting_pipeline_executor()  # pylint: disable=no-value-for-parameter
+    test_inform_selector.return_value = {}
+    test_inform_selector_and_trigger.side_effect = [(0, {}), (1, {}), (2, {}), (3, {}), (4, {}), (5, {})]
+
+    # each tuple is (key, timestamp, label)
+    first_batch = [
+        # trigger 0
+        (10, 1, 5),
+        (11, 2, 5),
+        # trigger 1 is empty
+        # trigger 2
+        (12, 3, 5),
+        (13, 4, 5),
+        # remaining data
+        (14, 5, 5),
+        (15, 6, 5),
+        (16, 7, 5),
+    ]
+    triggering_indices = [1, 1, 3]
+    test_get_number_of_samples.side_effect = [2, 0, 2]  # the size of trigger 0, 1, 2
+
+    pe._handle_triggers_within_batch(first_batch, triggering_indices)
+    assert pe.pipeline_log["supervisor"]["triggers"][0]["first_timestamp"] == 1
+    assert pe.pipeline_log["supervisor"]["triggers"][0]["last_timestamp"] == 2
+
+    assert "first_timestamp" not in pe.pipeline_log["supervisor"]["triggers"][1]
+    assert "last_timestamp" not in pe.pipeline_log["supervisor"]["triggers"][1]
+
+    assert pe.pipeline_log["supervisor"]["triggers"][2]["first_timestamp"] == 3
+    assert pe.pipeline_log["supervisor"]["triggers"][2]["last_timestamp"] == 4
+    assert pe.remaining_data_range == (5, 7)
```

Comment:
This looks good, I am just confused a bit with the logic it seems haha

Reply:
> How can triggering_data be empty if num_samples_in_trigger > 0? That confuses me a bit. And why can this only happen in the first trigger?

The first `triggering_data` in a batch is not the entire data within the corresponding trigger, but only its subset. The entire data can span over the boundary of two consecutive batches. In that case, it consists of the `remaining_data` from the previous batch and the first `triggering_data` in the current batch. An extreme case is that the part in the current batch (that is, `triggering_data`) is empty.

See the second time `_handle_triggers_within_batch` is called. The first trigger index is -1. So `triggering_data` is indeed `[]`, but the entire data also includes remaining data from the first batch, which is `[(14, 5, 5), (15, 6, 5), (16, 7, 5)]`.

As another example, trigger 5 consists of the first `triggering_data`, i.e. `[(20, 11, 5), (21, 12, 5), (22, 13, 5)]` in the third batch, and the `remaining data`, i.e. `[(19, 10, 5)]` from the second batch.

Reply:
The first paragraph completely cleared it up for me, thank you for the explanation. Makes total sense. Feel free to merge!